core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java

```diff
@@ -390,8 +390,7 @@ private void updateTaskStatus(TaskStatus taskStatus)
     private synchronized boolean isStageFlushing()
     {
         // to transition to flushing, there must be at least one flushing task, and all others must be flushing or finished.
-        return !flushingTasks.isEmpty()
-                && allTasks.stream().allMatch(taskId -> finishedTasks.contains(taskId) || flushingTasks.contains(taskId));
+        return !flushingTasks.isEmpty() && allTasks.size() == finishedTasks.size() + flushingTasks.size();
     }
 
     private synchronized boolean isStageFinished()
```

Comment:
Is this required to be synchronized?

Reply:
For now both `finishedTasks` and `allTasks` are marked as `@GuardedBy(this)` (and locked opportunistically). Potentially these don't need to be synchronzied. However, this would also mean that `isStageFinished` could return `true` even though `flushingTasks` is not yet empty (e.g. if there is concurrent `addFinishedTask` call).

I would rather not change it now.